LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/LittleEndianReader.cs

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Runtime.CompilerServices;
+
+    internal static class LittleEndianReader
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint ReadUInt24(ReadOnlySpan<byte> buffer) =>
+            buffer.Length >= 3 ? buffer[0] | ((uint)buffer[1] << 8) | ((uint)buffer[2] << 16) : ThrowInsufficientBufferLength();
+
+        [DoesNotReturn]
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Curious - why these `MethodImpl` attributes on these methods?

Reply:
To override JIT's heuristics. The JIT doesn't inline methods with a `throw` so the throwing is done in a separate method that it's told not to inline.

Suppose you have the following:

```c#
public class C
{
    public uint M(byte[] buffer) =>
        LittleEndianReader.ReadUInt24(buffer);
}
```

Then you can see the inlining in action (see also on [sharplab.io]):

```asm
C.M(Byte[])
    L0000: sub rsp, 0x28
    L0004: test rdx, rdx
    L0007: je short L001c
    L0009: lea rax, [rdx+0x10]
    L000d: mov edx, [rdx+8]
    L0010: cmp edx, 3
    L0013: jge short L0022
    L0015: call 0x00007ffb39690038
    L001a: jmp short L0037
    L001c: xor eax, eax
    L001e: xor edx, edx
    L0020: jmp short L0010
    L0022: movzx edx, byte ptr [rax]
    L0025: movzx ecx, byte ptr [rax+1]
    L0029: shl ecx, 8
    L002c: or edx, ecx
    L002e: movzx eax, byte ptr [rax+2]
    L0032: shl eax, 0x10
    L0035: or eax, edx
    L0037: add rsp, 0x28
    L003b: ret
```

Notice the length 3 check at offset L0010 and left-shifts (`shl`) at offsets L0029 and L0032 appear _within_ the compiled method.

If you remove the `[MethodImpl(MethodImplOptions.AggressiveInlining)]` decoration on `LittleEndianReader.ReadUInt24`, you get a shorted body and `call` instead on L0025:

```asm
C.M(Byte[])
    L0000: sub rsp, 0x38
    L0004: xor eax, eax
    L0006: mov [rsp+0x28], rax
    L000b: test rdx, rdx
    L000e: je short L0030
    L0010: lea rcx, [rdx+0x10]
    L0014: mov eax, [rdx+8]
    L0017: mov [rsp+0x28], rcx
    L001c: mov [rsp+0x30], eax
    L0020: lea rcx, [rsp+0x28]
    L0025: call 0x00007ffb39760030
    L002a: nop
    L002b: add rsp, 0x38
    L002f: ret
    L0030: xor ecx, ecx
    L0032: xor eax, eax
    L0034: jmp short L0017
```

[sharplab.io]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBEBLbAcwDsJcGDmFwsAwhAAmMAIK9sAGwCeuDrgDcNek1YAlAK68h+GOIj4ADhwUwoAZVsA3YTA00tAZiakGYmgG8aBmCmL30OIwYAWQAKYCUMGABtAF0GYH0AM0zbAEoGAF4APiCQsoAZDgwMGwBRXkkuXl0YbGkoFha2gFUASSNSFDisnKhczWoAX3daL2YkbwZK6rqGpq72gNLgpKiYDAALKV7LBRi9w+PTgHkLIQheURlublhcVUcYfoUIiO5clLbUI6BbhSIbPoDIYba68ZR2CzYXgAHniiSK6RGeUKGMMYAOMDAAGsYJIYkCyhlsrYWOUYLxuIcGEUCgwvAB+THUqBJOhpQgMGIxMEYXJU0ZJchpZHIhgADnyAqFIrFWJ5pGlsvISFyFLK+oNhpADAAKgcoBAAO79XAjYQcekYABCarpDMOMVy4xmZSSbAgrgAchAMC0MPooLxAdRfRcjpIThYznGrknbvdHixg99fgyAUCLFAOI5sIkQQwRabzVabXawA6jC7uW7GQdPTi9cFDhbLQxeDBezIoNx9CYjLUEGAYHcOA8YgAiWvZe2OrmjBg2d0HFjztAMefi2zz71TGhAA